### PR TITLE
Remove IndexSet overloads in DataOut and FEValues

### PR DESCRIPTION
**Pull request by masterleinad**
Part of #4387. The `IndexSet` instantiations for `DataOut` and `FEValues` are unintuitive and cause lots of compiler warnings with recent compilers. In particular, these instantiations are untested (apart from `step-41` and `step-42`) and it is not clear at all what the derivatives should represent.

### CODE REVIEW

**Comment by masterleinad** [comment]
Fixes the warnings in https://cdash.kyomu.43-1.org/viewBuildError.php?type=1&buildid=8508.

File: examples/step-42/step-42.cc [hunk 1]
```diff
@@ -1553,10 +1559,9 @@ namespace Step42
...
     fraction_of_plastic_q_points_per_cell = 0;
 
-    typename DoFHandler<dim>::active_cell_iterator cell =
-                                                     dof_handler.begin_active(),
-                                                   endc = dof_handler.end();
-    unsigned int cell_number                            = 0;
+    auto         cell        = dof_handler.begin_active();
+    const auto   endc        = dof_handler.end();
+    unsigned int cell_number = 0;
     for (; cell != endc; ++cell, ++cell_number)
```
Comment (bangerth) [inline examples/step-42/step-42.cc]: You can convert this into a range-based for loop if you get rid of the `cell_number` variable and instead use `cell->active_cell_index()`.

**Comment by drwells** [comment]
Looks good to me (aside from @bangerth's suggestions).

**Comment by masterleinad** [comment]
> Are there (internal) functions in the `fe_values.*` files you can get rid of now that you've removed the `IndexSet` instantiations?

No, I could not find more occurrences of "`IndexSet`" in the `fe_values.*` files.
I adressed your other comments in the last commits.

**Pull request merged**